File: common/src/main/java/cc/turtl/chiselmon/system/alert/AlertManager.java
package cc.turtl.chiselmon.system.alert;

import cc.turtl.chiselmon.ChiselmonConstants;
import cc.turtl.chiselmon.ChiselmonKeybinds;
import cc.turtl.chiselmon.api.PokemonEncounter;
import cc.turtl.chiselmon.api.Priority;
import cc.turtl.chiselmon.api.event.ChiselmonEvents;
import cc.turtl.chiselmon.api.filter.match.FilterMatchResult;
import cc.turtl.chiselmon.api.filter.match.FilterMatcher;
import cc.turtl.chiselmon.config.ChiselmonConfig;
import cc.turtl.chiselmon.config.category.AlertConfig;
import cc.turtl.chiselmon.system.alert.action.*;
import cc.turtl.chiselmon.system.tracker.TrackerManager;
import cc.turtl.chiselmon.util.MessageUtils;
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;

import java.util.*;

public class AlertManager {
    private static final int SOUND_DELAY_TICKS = 20;

    private static final AlertManager INSTANCE = new AlertManager();

    private final List<AlertAction> oneTimeActions = List.of(new MessageAction(), new SoundAction(), new DiscordAction());
    private final List<AlertAction> continuousActions = List.of(new GlowAction());
    private final SoundAction repeatingSoundAction = new SoundAction();

    private Set<UUID> mutedUuids;
    private Set<UUID> actionedUuids;
    private int soundDelayRemaining = 0;
    private boolean active = false;

    private AlertManager() {
    }

    public static AlertManager getInstance() {
        return INSTANCE;
    }

    public void init() {
        ChiselmonEvents.LEVEL_CONNECTED.subscribe(Priority.LOW, e -> onWorldJoin());
        ChiselmonEvents.LEVEL_DISCONNECTED.subscribe(Priority.LOW, e -> onWorldLeave());
        // this is low priority for the glow logic that might clash with despawn glow. i want alert to override despawn glow
        ChiselmonEvents.CLIENT_POST_TICK.subscribe(Priority.LOW, e -> {
            if (active) tick();
        });
        ChiselmonConstants.LOGGER.info("AlertSystem initialized");
    }

    private void onWorldJoin() {
        if (active) {
            ChiselmonConstants.LOGGER.warn("New world joined before AlertSystem was disposed - resetting");
        }
        mutedUuids = new HashSet<>();
        actionedUuids = new HashSet<>();
        soundDelayRemaining = 0;
        active = true;
        ChiselmonConstants.LOGGER.debug("AlertSystem started");
    }

    private void onWorldLeave() {
        mutedUuids = null;
        actionedUuids = null;
        soundDelayRemaining = 0;
        active = false;
        ChiselmonConstants.LOGGER.debug("AlertSystem disposed");
    }

    private void tick() {
        AlertConfig config = ChiselmonConfig.get().alert;
        if (!config.masterEnabled) return;

        while (ChiselmonKeybinds.MUTE_ALERTS.consumeClick()) {
            LocalPlayer player = Minecraft.getInstance().player;
            if (player != null) {
                muteAll();
                MessageUtils.sendSuccess(player, "All active alert muted");
            }
        }

        // Track the "best" filter match for the sound this tick
        AlertContext bestSoundContext = null;

        for (PokemonEntity pe : getLoadedPokemonEntities()) {
            UUID uuid = pe.getUUID();
            if (!pe.getBusyLocks().isEmpty()) mute(uuid);

            FilterMatchResult result = FilterMatcher.match(pe.getPokemon());
            if (result.allMatches().isEmpty()) continue;

            AlertContext ctx = new AlertContext(pe, result.allMatches(), isMuted(uuid), config, PokemonEncounter.from(pe));

            continuousActions.forEach(action -> action.execute(ctx));

            if (!actionedUuids.contains(uuid)) {
                oneTimeActions.forEach(action -> action.execute(ctx));
                actionedUuids.add(uuid);
            }

            // Update the best sound if the pokemon isn't muted and their filter ctx has a higher priority than the current best
            if (ctx.shouldRepeatingSound()) {
                if (bestSoundContext == null || ctx.soundFilter().priority().isHigherThan(bestSoundContext.soundFilter().priority())) {
                    bestSoundContext = ctx;
                }
            }
        }

        // replay the sound action for repeating sound alert
        if (soundDelayRemaining > 0) {
            soundDelayRemaining--;
        } else if (bestSoundContext != null) {
            repeatingSoundAction.executeRepeating(bestSoundContext);
            soundDelayRemaining = SOUND_DELAY_TICKS;
        }
    }

    private Collection<PokemonEntity> getLoadedPokemonEntities() {
        return TrackerManager.getInstance().getTracker().getCurrentlyLoaded().values();
    }

    public void mute(UUID uuid) {
        mutedUuids.add(uuid);
    }

    public void muteAll() {
        mutedUuids.addAll(TrackerManager.getInstance().getTracker().getCurrentlyLoaded().keySet());
    }

    public void unmuteAll() {
        mutedUuids.clear();
    }

    public boolean isMuted(UUID uuid) {
        return mutedUuids.contains(uuid);
    }

    public Set<UUID> getMutedUuids() {
        return mutedUuids;
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/system/alert/AlertManager.kt
package cc.turtl.chiselmon.system.alert

import cc.turtl.chiselmon.client.ChiselmonKeybindsKt
import cc.turtl.chiselmon.client.config.ChiselmonConfig
import cc.turtl.chiselmon.client.util.sendSuccess
import cc.turtl.chiselmon.core.ChiselmonConstants
import cc.turtl.chiselmon.core.api.PokemonEncounter
import cc.turtl.chiselmon.core.api.filter.match.FilterMatcher
import cc.turtl.chiselmon.system.alert.action.*
import cc.turtl.chiselmon.system.tracker.TrackerManager
import cc.turtl.turtlshell.api.client.ClientEvents
import net.minecraft.client.Minecraft
import java.util.*

object AlertManager {

    private const val SOUND_DELAY_TICKS = 20

    private val oneTimeActions: List<AlertAction> = listOf(MessageAction(), SoundAction(), DiscordAction())
    private val continuousActions: List<AlertAction> = listOf(GlowAction())
    private val repeatingSoundAction = SoundAction()

    private val mutedUuids = hashSetOf<UUID>()
    private val actionedUuids = hashSetOf<UUID>()
    private var soundDelayRemaining = 0
    private var active = false

    fun init() {
        ClientEvents.LEVEL_CONNECTED.subscribe { onWorldJoin() }
        ClientEvents.LEVEL_DISCONNECTED.subscribe { onWorldLeave() }
        // low priority for the glow logic that might clash with despawn glow -- alert should override despawn glow
        ClientEvents.TICK_POST.subscribe { if (active) tick() }
        ChiselmonConstants.LOGGER.info("AlertSystem initialized")
    }

    private fun reset() {
        mutedUuids.clear()
        actionedUuids.clear()
        soundDelayRemaining = 0
    }

    private fun onWorldJoin() {
        if (active) {
            ChiselmonConstants.LOGGER.warn("New world joined before AlertSystem was disposed - resetting")
        }
        reset()
        active = true
        ChiselmonConstants.LOGGER.debug("AlertSystem started")
    }

    private fun onWorldLeave() {
        reset()
        active = false
        ChiselmonConstants.LOGGER.debug("AlertSystem disposed")
    }

    private fun tick() {
        val config = ChiselmonConfig.alert
        if (!config.masterEnabled) return

        while (ChiselmonKeybindsKt.MUTE_ALERTS.consumeClick()) {
            Minecraft.getInstance().player?.let { player ->
                muteAll()
                sendSuccess(player, "All active alert muted")
            }
        }

        // Track the "best" filter match for the sound this tick
        var bestSoundContext: AlertContext? = null

        for (pe in TrackerManager.tracker.currentlyLoaded.values) {
            val uuid = pe.uuid
            if (pe.busyLocks.isNotEmpty()) mute(uuid)

            val result = FilterMatcher.match(pe.pokemon)
            if (result.allMatches.isEmpty()) continue

            val ctx = AlertContext(
                entity = pe,
                filters = result.allMatches,
                isMuted = isMuted(uuid),
                config = config,
                encounter = PokemonEncounter.from(pe),
            )

            continuousActions.forEach { it.execute(ctx) }

            if (actionedUuids.add(uuid)) {
                oneTimeActions.forEach { it.execute(ctx) }
            }

            // Promote ctx to the best sound candidate if its filter out-prioritises the current best
            if (ctx.shouldRepeatingSound) {
                val currentBestPriority = bestSoundContext?.soundFilter?.priority
                val ctxPriority = ctx.soundFilter?.priority
                if (currentBestPriority == null || (ctxPriority != null && ctxPriority.isHigherThan(currentBestPriority))) {
                    bestSoundContext = ctx
                }
            }
        }

        // Replay the sound action for repeating sound alerts on a delay
        if (soundDelayRemaining > 0) {
            soundDelayRemaining--
        } else bestSoundContext?.let {
            repeatingSoundAction.executeRepeating(it)
            soundDelayRemaining = SOUND_DELAY_TICKS
        }
    }

    fun mute(uuid: UUID) = mutedUuids.add(uuid)

    fun muteAll() = mutedUuids.addAll(TrackerManager.tracker.currentlyLoaded.keys)

    fun unmuteAll() = mutedUuids.clear()

    fun isMuted(uuid: UUID): Boolean = uuid in mutedUuids

    fun getMutedUuids(): Set<UUID> = mutedUuids
}
